src/librustc_mir/interpret/operator.rs

```diff
@@ -153,11 +85,19 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
 
         // These ops can have an RHS with a different numeric type.
         if right_kind.is_int() && (bin_op == Shl || bin_op == Shr) {
-            return match bin_op {
-                Shl => int_shift!(left_kind, overflowing_shl, l, r),
-                Shr => int_shift!(left_kind, overflowing_shr, l, r),
+            let op: fn(u128, u32) -> (u128, bool) = match bin_op {
+                Shl => u128::overflowing_shl,
+                Shr => u128::overflowing_shr,
                 _ => bug!("it has already been checked that this is a shift op"),
             };
+            let l = if left_ty.is_signed() {
+                self.sign_extend(l, left_ty)?
+            } else {
+                l
+            };
+            let (result, oflo) = op(l, r as u32);
+            let truncated = self.truncate(result, left_ty)?;
+            return Ok((PrimVal::Bytes(truncated), oflo || truncated != result));
```

Comment:
I don't think this works for signed shifts (since the truncated form doesn't have the sign-extend bits anymore) - also, doesn't "overflowing" refer to the shift count being larger the bitwidth, not the value going out of range? but I'm not sure.

Reply:
yea I already fixed that ^^